r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -1363,40 +1325,12 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> FieldGadget<SimulationF,
         mut cs: CS,
         fe: &SimulationF,
     ) -> Result<Self, SynthesisError> {
-
-        if super::is_pseudo_mersenne::<SimulationF>() {
-            let other = NonNativeFieldGadget::from_value(cs.ns(|| "alloc constant for mul"), fe);
-            return self.mul_for_pseudomersenne(&mut cs, &other, true);
+        let other = NonNativeFieldGadget::from_value(cs.ns(|| "alloc constant for mul"), fe);
+        return if super::is_pseudo_mersenne::<SimulationF>() {
+            self.mul_for_pseudomersenne(&mut cs, &other, true)
+        } else {
+            self.mul_no_pseudomersenne(&mut cs, &other, true)
```

Comment:
Can't you just call mul() ?

Reply:
These two private functions are the actual implementation of mul employing the optimized algorithm for pseudo-mersenne fields and the non optimized algorithm, respectively. Mul now is just a dispatcher that calls either of these two functions depending on `SimulationF`